_posts/sumtyper.md

```diff
@@ -0,0 +1,406 @@
+
+For eksempel, om du lager en ny klasse med tre felter av typen boolean, er antall mulige verdier 2 × 2 × 2 = 8. Som vi så i eksempelet over, blir dette fort uhåndterlig, og vi ender opp med mange muligheter som ikke finnes i den virkelige verden.
+
+En annen måte å konstruere nye typer på kalles sumtyper. I forskjellige programmeringsspråk kalles de også tagged unions, disjoint unions og variants. Forskjellen fra produkttyper er at de består av en gitt mengde varianter, og antall mulige verdier av denne nye typen er summen av varianter. Det betyr at om vi legger til en ny variant, blir det bare en ny mulig verdi. Dette er styrken til sumtyper, og grunnen til at vi kan unngå mye av den uheldige kompleksiteten vi så tidligere i adresseeksempelet vårt.
+
+Aller helst vil vi også skrive sumtypene på en typesikker måte, slik at vi får en kompilatorfeil om vi ikke håndterer alle variantene. Det ville forhindret feilen i eksempelet vårt, hvor vi ikke håndterer matrikkeladressen. Det vil også sørge for at vi blir nødt til å håndtere nye varianter som blir innført senere.
+
+I eksempelet vårt burde Adresse modelleres som en sumtype med to varianter: en vegadresse og en matrikkeladresse. Men hvordan gjør man det i Java?
+
+## Første forsøk på omskriving av adresseklassen
+
+Det første vi kan forsøke er å la Adresse være en abstrakt klasse, og hver variant være en konkret implementasjon. Vi kan også la funksjonen adresselapp være en abstrakt metode på Adresse, og implementere den i hver av subklassene.
+
+```java
+public abstract class Adresse {
+
+  String navn;
+
+  public abstract String adresselapp();
+}
+
+class Vegadresse extends Adresse {
+
+  String veiadresse;
+  String postnummer;
+  String poststed;
+}
+
+class MatrikkelAdresse extends Adresse {
+
+  Integer gårdsnummer;
+  Integer bruksnummer;
+}
+
+```
+
+Dette løser et par av problemene vi hadde med den første implementasjonen. Det er tydelig at navn er felles for begge variantene, og at de andre feltene er spesifikke for hver sin variant. Metoden adresselapp er en abstrakt metode, så alle subklassene må implementere den. Hvis vi innfører en ny subklasse, vil kompilatoren si fra hvis denne metoden ikke er implementert.
+
+Dette fungerer bra i små prosjekter, hvor vi har klassene våre i en modul, og vi har rimelig god oversikt over domenemodellen. I større prosjekter fungerer det ikke så godt lenger.
+
+Vi kan tenke oss at en sentral klasse som Adresse vil bli brukt mange steder i systemet. Ofte er den ikke definert i modulen vi jobber med, men den er inkludert som en avhengighet til en annen modul. Kanskje er det mange andre moduler på prosjektet som avhenger av den samme klassen, og bruker den til forskjellige ting. Hvis alle som jobber med Adresse utvider klassen med sine egne metoder, blir det fort uoversiktlig. Det er mye bedre om Adresse bare er en dataklasse, altså en klasse som bare holder på data. Så kan alle skrive sine egne funksjoner som tar adresseobjekter som parametere.
```

Comment:
> Det er mye bedre om Adresse bare er en dataklasse

Kan man utbrodere dette litt? Det er her jeg føler man kan miste noen av de som ikke er interessert i FP fra før. For objektorientert-entusiaster er denne implementasjonen i "Første forsøk" slik man løser disse situasjonene, og jeg tror det er rundt her i bloggposten "kneika" er for å få overbevist de om verdien med å bruke visitor pattern eller church encoding.